### PR TITLE
Sk/add lookup

### PR DESCRIPTION
**Pull request by sriharshakandala**
Lookup tables are added. 
Gas optics and cloud optics support added. 
Documentation is updated.
Supports GPU acceleration on CUDA devices using native CUDA kernels.
Uses native Julia native multithreading on CPU.

### CODE REVIEW

File: src/optics/RTE.jl [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    as::AS         # atmospheric state
+    op::OP         # optical properties
+    src_lw::SL     # source functions
+    src_sw::SS     # source functions
+    bcs_lw::BCL    # boundary conditions
+    bcs_sw::BCS    # boundary conditions
+    fluxb_lw::FXBL # temporay storage for bandwise calculations
+    fluxb_sw::FXBS # temporay storage for bandwise calculations
+    flux_lw::FXL   # fluxes for longwave problem
+    flux_sw::FXS   # fluxes for longwave problem
```
Comment (szy21) [inline src/optics/RTE.jl:67]:
```suggestion
    flux_sw::FXS   # fluxes for showrtwave problem
```

File: src/optics/RTE.jl [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    FXL<:Union{FluxLW{FT,FTA2D},Nothing},
+    FXS<:Union{FluxSW{FT,FTA2D},Nothing},
+}
+    as::AS         # atmospheric state
+    op::OP         # optical properties
+    src_lw::SL     # source functions
+    src_sw::SS     # source functions
+    bcs_lw::BCL    # boundary conditions
+    bcs_sw::BCS    # boundary conditions
+    fluxb_lw::FXBL # temporay storage for bandwise calculations
```
Comment (szy21) [inline src/optics/RTE.jl:64]:
```suggestion
    fluxb_lw::FXBL # temporary storage for bandwise calculations
```

File: src/optics/RTE.jl [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    fluxb_sw::FXBS # temporay storage for bandwise calculations
```
Comment (szy21) [inline src/optics/RTE.jl:65]:
```suggestion
    fluxb_sw::FXBS # temporary storage for bandwise calculations
```

File: src/optics/Sources.jl [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    FTA1D<:AbstractArray{FT,1},
+    FTA2D<:AbstractArray{FT,2},
+} <: AbstractSourceLW{FT,FTA1D,FTA2D}
+    "Planck source at layer average temperature `[W/m2]` `(nlay, ncol)`"
+    lay_source::FTA2D
+    "Planck source at layer edge in increasing ilay direction `[W/m2]` `(nlay, ncol)`, includes spectral weighting that accounts for state-dependent frequency to g-space mapping"
+    lev_source_inc::FTA2D
+    "Planck source at layer edge in decreasing ilay direction `[W/m2]` `(nlay, ncol)`, includes spectral weighting that accounts for state-dependent frequency to g-space mapping"
+    lev_source_dec::FTA2D
+    "Surface source `(ncol)`"
```
Comment (szy21) [inline src/optics/Sources.jl:41]:
```suggestion
    "Surface source `[W/m2]` `(ncol)`"
```

File: src/optics/Sources.jl [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
```
Comment (szy21) [inline src/optics/Sources.jl:89]:
```suggestion
    "Surface source `[W/m2]` `(ncol)`"
```

File: src/rte/RTESolverKernels.jl [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+    # Albedo of lowest level is the surface albedo...
+    @inbounds albedo[1, gcol] = FT(1) - sfc_emis[ibnd, gcol]
+    # ... and source of diffuse radiation is surface emission
+    @inbounds src[1, gcol] = FT(π) * sfc_emis[ibnd, gcol] * sfc_source[gcol]
+    #--------------------------------------------------------------------------------------
+    #    # Albedo of lowest level is the surface albedo...
+    #    albedo[1, gcol] = FT(1) - sfc_emis[gcol]
+    #    # ... and source of diffuse radiation is surface emission
+    #    src[1, gcol] = FT(π) * sfc_emis[gcol] * sfc_source[gcol]
+    #--------------------------------------------------------
```
Comment (szy21) [inline src/rte/RTESolverKernels.jl:277]: Is it ok to remove this?

File: Project.toml [hunk 7]
```diff
@@ -7,12 +7,28 @@ version = "0.1.0"
 Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
 CLIMAParameters = "6eacf6c3-8458-43b9-ae03-caf5306d3d53"
 CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
+Compat = "34da2185-b29b-5c13-b0c7-acf172513d20"
+Coverage = "a2441757-f6aa-5fb2-8edb-039e3f45d037"
 DocStringExtensions = "ffbed154-4ef7-542d-bbb7-c09d3a79fcae"
+DocumenterMarkdown = "997ab1e6-3595-5248-9280-8efb232c3433"
+DocumenterTools = "35a29f4d-8980-5a13-9543-d66fff28ecb8"
```
Comment (simonbyrne) [inline Project.toml:14]: These shouldn't be needed

File: Project.toml [hunk 8]
```diff
@@ -7,12 +7,28 @@ version = "0.1.0"
...
 GaussQuadrature = "d54b0c1a-921d-58e0-8e36-89d8069c0969"
-KernelAbstractions = "63c18a36-062a-441e-b654-da1e3ab1ce7c"
+InteractiveUtils = "b77e0a4c-d291-57a0-90e8-8db25a27a240"
```
Comment (simonbyrne) [inline Project.toml:16]: Or this: go through and check which are actually used.

File: src/rte/RTESolverKernels.jl [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+    # Albedo of lowest level is the surface albedo...
+    @inbounds albedo[1, gcol] = bcs_sw.sfc_alb_diffuse[ibnd, gcol]
+    # ... and source of diffuse radiation is surface emission
+    @inbounds src[1, gcol] = sfc_source[gcol]
+    #--------------------------------------------------------------------------------------
+    #    # Albedo of lowest level is the surface albedo...
+    #    albedo[1, gcol] = FT(1) - sfc_emis[gcol]
+    #    # ... and source of diffuse radiation is surface emission
+    #    src[1, gcol] = FT(π) * sfc_emis[gcol] * sfc_source[gcol]
+    #--------------------------------------------------------
```
Comment (szy21) [inline src/rte/RTESolverKernels.jl:527]: Is it ok to remove this?

File: Project.toml [hunk 10]
```diff
@@ -7,12 +7,28 @@ version = "0.1.0"
...
+NCDatasets = "85f8d34a-cbdd-5861-8df4-14fed0d494ab"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
+Profile = "9abbd945-dff8-562f-b5e8-e1ebf5ef1b79"
+ProgressMeter = "92933f4c-e287-5a05-a399-4b506db050ca"
 StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
Comment (simonbyrne) [inline Project.toml:22]: Test can be a test only dependency, see http://pkgdocs.julialang.org/v1/creating-packages/#Test-specific-dependencies-in-Julia-1.2-and-above

File: docs/make.jl [hunk 11]
```diff
@@ -13,28 +16,24 @@ makedocs(
             ),
         )),
     ),
-    clean = true,
-    modules = [RRTMGP],
+    clean = false,
+    modules = [Documenter, RRTMGP],
```
Comment (simonbyrne) [inline docs/make.jl:20]:
```suggestion
    clean = true,
    modules = [RRTMGP],
```

**Comment by simonbyrne** [comment]
@dennisYatunin can you have a shot at reviewing this?

File: test/gray_atm.jl [hunk 12]
```diff
@@ -24,84 +27,114 @@ Example program to demonstrate the calculation of longwave radiative fluxes in a
...
         lat = DA{FT}(range(FT(-π / 2), FT(π / 2), length = ncol)) # latitude
     end
 
-    println("Running longwave test for gray atmosphere model - $(OPC); ncol = $ncol; DA = $DA")
-    gray_flux = GrayFluxLW(ncol, nlay, nlev, FT, DA)
+    as = setup_gray_as_pr_grid(nlay, lat, p0, pe, param_set, DA)
+    op = OPC(FT, ncol, nlay, DA)
+    src_lw = source_func_longwave(FT, ncol, nlay, opc, DA)
+    bcs_lw = LwBCs(DA{FT,2}(sfc_emis), inc_flux)
+    fluxb_lw = nothing #FluxLW(ncol, nlay, FT, DA)
```
Comment (szy21) [inline test/gray_atm.jl:64]: Is it ok to remove the comment here?

Comment (sriharshakandala) [reply test/gray_atm.jl]: Done.

**Comment by szy21** [comment]
> @dennisYatunin can you have a shot at reviewing this?

I just quickly went through the code, but yes @dennisYatunin it would be great if you can look at the code in more detail, so I can focus on the documentation.

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Review by szy21 — APPROVED** [review]

**Pull request merged**